### DISCUSSION TITLE
Add option to preview media in rich text interfaces using current user's permissions

### DISCUSSION THREAD

**Discussion by Thermal-Shock** [directus/directus#15936] [Feature Requests]
### Describe the Bug

The Image field (file-image) does not require an access token to see the image in the interface. The same should be true of the preview mode of the markdown (input-rich-text-md) and WYSIWYG (input-rich-text-html) interfaces. 

The reason this works for the image field is because the code transparently appends the result of calling app/src/api.ts-addTokenToURL() to the URL, however the others only do this based on a static access token prop.

This is a bug in the sense that there is a lack of integration between Directus and codemirror lang-html and lang-markdown. 

This can be resolved by modifying the preview in codemirror to add an opportunity for Directus to preprocess the HTML prior to rendering by detecting any img /asset/ URLs, and append the Directus addTokenToURL() transparently. 

### To Reproduce

Add an image from the file library and note that it displays just fine. Note that the field interface does not require an access token. 

Add a markdown or WYSIWYG interface and add an image from the file library, and click preview. The image will not preview and the directus log will show a GET of the asset with a 403 (permission) error. 

Generating and adding a static access token in both the user and the field interface resolves the issue, but then all the URLs have access_token=... appended to them. This is problematic and could be handled transparently.

**Comment by azrikahar** [comment]
I believe one use case that does require `access_token` to be appended directly to the assets in markdown or WYSIWYG interfaces is that if they are being used externally (such as a blog post, article etc). 🤔 Though understandably the same issue can be said for image field that is shown externally as well.

**Reply by Thermal-Shock** [reply]
Thank you for your reply. I'm disappointed this was converted into a discussion. 

I still see this as an issue since these interfaces are is not consistent with the other Directus preview interfaces that append the API token automatically since they are previewing within the app. It just doesn't seem appropriate to have to use an access token prop and litter otherwise neutral markdown with an access token that I also might not want to be known. 

This works for the image interface component and I also include the EditorJS component that also append the token to the URL for the interface. Naturally they both allow you to select an asset the their URLs start with "/asset/". There are other vue components that preview assets as well that do the same thing.

```
api.defaults?.headers?.['Authorization']?.split(' ')[1] ||
api.defaults?.headers?.common?.['Authorization']?.split(' ')[1] || null
```

The same should (or could) be true of both of the rich text (markdown and WYSIWYG) interfaces or at just have a toggle setting in their interface components to append token to asset URLs prior to preview. By token I mean the above code.

If I was to create my own version of the markdown and WYSIWYG interface, is there a hook in codemirror to modify the HTML prior to preview so I can do this replacement myself?

**Comment by rijkvanzanten** [comment]
While I agree there's room for additional improvements in options for these interfaces, I don't quite agree with the _"This is a bug in the sense that there is a lack of integration between Directus and codemirror lang-html and lang-markdown."_ callout. The WYSIWYG interface is meant to be "what you see is what you get", so it will render the preview as it's saved to the database. This also means that it'll preview the images with whatever the permissions of the included token are (or public if omitted), as this is one-to-one what you would get when rendering this HTML in another place.

> It just doesn't seem appropriate to have to use an access token prop and litter otherwise neutral markdown with an access token that I also might not want to be known.

This is a valid statement, but should be resolved by tweaking the read permissions for the public role to the file, rather than having the wysiwyg interface fake the preview, and then having you deal with the aftermath of that in your own usage of the stored html 🤔

**Reply by Thermal-Shock** [reply]
Thanks @rijkvanzanten. 

The saved HTML would not be any different. It would simply be a modification to the preview in the interface to optionally preprocess the HTML to detect "/asset/" URLs and append the API token prior to displaying the final HTML in the DOM.

Without this modification, you cannot preview asset images without making them public to some degree. It is not necessary to make images public when using the file-image interface and within the app. 

Agree to disagree though I suppose. 

I'll work on sectioning off the images into folders and then modify the public read permission by adding an item permission filter to only allow those folders to be publicly accessible.

**Reply by rijkvanzanten** [reply]
> It would simply be a modification to the preview in the interface to optionally preprocess the HTML to detect "/asset/" URLs and append the API token prior to displaying the final HTML in the DOM.

Yeah! And I'm saying that would be confusing, as that preview would be different from what you'd see when you render that HTML in your own project or app 🙂 What you're describing above in your OP is actually how it used to work before, and it was later changed as a lot of people got confused as to why their HTML worked when previewed in Directus, but not when using it on their own website

**Reply by Thermal-Shock** [reply]
I respect what you are saying but I feel there should at least be an "Append access token to assets in preview" option instead of specifying a static token or changing public permissions. However I get that this may not be trivial due to possibly lack of integration with codemirror or that it does not fit within the design philosophy of Directus. 

Thanks for listening and your discussion. I appreciate it.

**Reply by rijkvanzanten** [reply]
> but I feel there should at least be an "Append access token to assets in preview" option

Great suggestion. Lets use this discussion as the overall feature request for that 🙂 

>  **_instead of_** specifying a static token or changing public permissions 

(emphasis mine) I don't think we should flat out replace it. You still need to have a way to inject the tokens if you want to, as that might be how you're intending on doing access control from the HTML (we can't know that). For the previewing piece, yes, it'd be very useful to toggle between "Preview what's saved" and "Preview using the current user's permissions" 🙂 

> this may not be trivial due to possibly lack of integration with codemirror

True, though that hasn't stopped us in the past 😁

**Comment by piotr-cz** [comment]
To rephrase, this issue prevents non-public images to be shown in editor preview.

A workaround is to set Static Access Token in Text field interface - however then it's being appended to URLs and saved along markdown/ html output.
In effect whenever Static Access Token is regenerated, all image URLs must be updated.

I agree with @Thermal-Shock that this is rather bug than feature as Image field shows previews with attaching current users' JWT both to image URL and Authorization header.

The _marked_ package has an option to [extend native Renderer](https://marked.js.org/using_pro#renderer)'s image where current users' JWT may be added to image URL in editor preview mode when URL starts with API Assets string.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
